### PR TITLE
Voting rounds

### PR DESCRIPTION
**Pull request by rao-kartik**

### CODE REVIEW

File: Week-3/4_Voting/OneVoteRound.sol [hunk 1]
```diff
@@ -0,0 +1,92 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.16;
+
+contract oneVotingRound {
+    address admin;
```
Comment (mahima-manik) [inline Week-3/4_Voting/OneVoteRound.sol]: Make it immutable - no one should be able to change the admin once the contract is deployed

File: Week-3/4_Voting/OneVoteRound.sol [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    // adding contract deployer as admin
+    constructor() {
+        admin = msg.sender;
+    }
+
+    function registerCandidate() external checkVotingStarted checkVotingStopped("You are late. Voting is completed.") {
+        // checking if candidate is registered or not
+        require(CandidatesRecord[msg.sender].addr == address(0), "You are already registered.");
```
Comment (mahima-manik) [inline Week-3/4_Voting/OneVoteRound.sol]: Can you use bool 'isRegistered' in struct to check if the candidate is registered? 

Reason: The storage cost for an address type is 20 bytes, while the storage cost for a bool type is 1 byte

File: Week-3/4_Voting/OneVoteRound.sol [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    uint256 leadingCandidateVotes;
+
+    struct Candidate {
+        address addr;
+        uint256 votes;
+    }
+
+
+    mapping(address => Candidate) CandidatesRecord; // to check if candidate is registered
+    mapping(address => bool) votersList; // to check if voters have voted or not
```
Comment (mahima-manik) [inline Week-3/4_Voting/OneVoteRound.sol]: `votersList` name does not convey what it stores. Consider naming it to: hasVoted or something else.

File: Week-3/4_Voting/OneVoteRound.sol [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        require(!votersList[msg.sender], "You have already voted");
+
+        // setting true if voted
+        votersList[msg.sender] = true;
+
+        // vote
+        uint _totalVotes = CandidatesRecord[_voteTo].votes + 1;
+        CandidatesRecord[_voteTo].votes = _totalVotes;
+
+        if(_totalVotes > leadingCandidateVotes){
```
Comment (mahima-manik) [inline Week-3/4_Voting/OneVoteRound.sol]: Directly increment CandidatesRecord[_voteTo].votes and check on line 87. We do not need to create new variable `_totalVotes` for the same

File: Week-3/4_Voting/Readme.md [hunk 5]
```diff
@@ -0,0 +1,24 @@
+# Write 2 Voting Contracts
+
+## One voting round
+1. Voting started and stopped by admin
+2. Candidates should register themselves before voting starts
+3. Each voter able to vote only once
```
Comment (mahima-manik) [inline Week-3/4_Voting/Readme.md:6]: Add: Voters should be able to vote for registered candidated

**Review by mahima-manik — CHANGES_REQUESTED** [review]

File: Week-3/4_Voting/OneVoteRound.sol (hunk 1 above)
Comment (rao-kartik) [reply Week-3/4_Voting/OneVoteRound.sol]: Done

File: Week-3/4_Voting/OneVoteRound.sol (hunk 2 above)
Comment (rao-kartik) [reply Week-3/4_Voting/OneVoteRound.sol]: Done

File: Week-3/4_Voting/OneVoteRound.sol (hunk 3 above)
Comment (rao-kartik) [reply Week-3/4_Voting/OneVoteRound.sol]: Done

File: Week-3/4_Voting/OneVoteRound.sol (hunk 4 above)
Comment (rao-kartik) [reply Week-3/4_Voting/OneVoteRound.sol]: done

File: Week-3/4_Voting/Readme.md (hunk 5 above)
Comment (rao-kartik) [reply Week-3/4_Voting/Readme.md:6]: done

**Review by SrikanthAlva — APPROVED** [review]

**Pull request merged**